### PR TITLE
Remove partition columns before writing partitioned file

### PR DESCRIPTION
**Pull request by JkSelf**
Spark distinguishes between [data columns](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L583
) and [partition columns](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L584), and it only writes the [data columns](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L298) into the file. However, Velox writes both the data columns and partition columns into the file. This PR removes the partition columns from the row vector before writing file.

### CODE REVIEW

**Comment by JkSelf** [comment]
@mbasmanova Can you help to review? Thanks.

**Review by mbasmanova — COMMENTED** [review]
CC: @gggrace14 

I wonder if this is a bug in the query planner. Ge, would you help take a look?

**Comment by gggrace14** [comment]
> CC: @gggrace14
> 
> I wonder if this is a bug in the query planner. Ge, would you help take a look?

Looking at it

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp]: It looks like a repeated pattern that you call filterPartitionColumns() before calling write(index, dataInput). Does it make sense if we put filterPartitionColumns() into the body of write()? It looks to me your filterPartitionColumns() also works well when getPartitionChannels() is empty vector.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:424]: No need particular for PARQUET format. I think the change should apply to all formats we support so far.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:78]: For the 4th param, input->pool() is better than passed param to this func.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:46]: `std::vector<TypePtr>`

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [reply velox/connectors/hive/HiveDataSink.cpp:78]: For the 2nd param, would it be okay to pass std::move() to a param of type `const std::vector<VectorPtr>&`?

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:53]: For the 3rd param, it should be better to be input->nulls().

Here I find a good example of how to make a RowVector from a subset of child vectors.
https://github.com/facebookincubator/velox/blob/2dc97deebba2664c4620f1e462476c5692691e4d/velox/exec/TableWriter.cpp#L129-L135

It looks to me more clean to not have this helper func makeRowVector() but directly call `std::make_shared<RowVector>` in the body of filterPartitionColumns().

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:56]: Does it sound better to call this func makeDataInput() or filterOutPartitionColumns()?

**Comment by gggrace14** [comment]
> CC: @gggrace14
> 
> I wonder if this is a bug in the query planner. Ge, would you help take a look?

Hi @mbasmanova  Masha, I checked the Presto code, and it does look like a bug in the original today. That is, Presto does not include partition columns when writing to a file under a partition directory today. So I think we should have this change (and not for only parquet file).

Also I will run Presto Verifier after or before this change lands, as it is on a critical path.

Better to ask you to also take a look to help confirm.

https://github.com/prestodb/presto/blob/9104d718332913be6dc1f461917cd1b8594b04ec/presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java#L390
https://github.com/prestodb/presto/blob/9104d718332913be6dc1f461917cd1b8594b04ec/presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java#L509-L517
https://github.com/prestodb/presto/blob/9104d718332913be6dc1f461917cd1b8594b04ec/presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java#L163

**Comment by mbasmanova** [comment]
@gggrace14 Ge, thank you for looking and commenting. Sounds like this is a generic bug not specific to Parquet. Indeed, partition keys should not be written in to files, not only for Parquet, but for all formats.

@JkSelf This is a nice finding. Let's update PR description to explain that this is a generic bug. You may mention that it was discovered by running Spark tests on Gluten. I'll take a look at the code. Thanks.

File: velox/exec/tests/TableWriteTest.cpp
Comment (mbasmanova) [inline velox/exec/tests/TableWriteTest.cpp:1079]: Remove commented out code.

File: velox/exec/tests/TableWriteTest.cpp
Comment (mbasmanova) [inline velox/exec/tests/TableWriteTest.cpp:1991]: Do we need a new test or could we modify existing tests to verify the list of columns written into the file? I assume there are existing tests that write partitioned tables. These test would start failing once we add verification for list of columns in the file, no?

File: velox/connectors/hive/HiveDataSink.cpp
Comment (mbasmanova) [inline velox/connectors/hive/HiveDataSink.cpp:423]: This code repeats. Let's refactor to avoid copy-paste.

I suggest the following:

- In the constructor, figure out which input columns need to be written and store indices of these in a member variable.
- Here, right after "Lazy load all the input columns." loop, create a vector with just the necessary columns from input. Use this vector in calls to write.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (mbasmanova) [inline velox/connectors/hive/HiveDataSink.cpp:77]: Remove makeRowVector function and create RowVector directly.

**Review by mbasmanova — COMMENTED** [review]
@JkSelf Thank you for identifying this issue and working on a fix. Some comments below.

File: velox/exec/tests/TableWriteTest.cpp
Comment (JkSelf) [reply velox/exec/tests/TableWriteTest.cpp]: @mbasmanova  Removed.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (JkSelf) [reply velox/connectors/hive/HiveDataSink.cpp]: @mbasmanova Good suggestions. I have updated.

**Comment by JkSelf** [comment]
@mbasmanova @gggrace14 Thanks for your review. I have updated all your comments. Can you help to review again? Thanks.

File: velox/connectors/hive/HiveDataSink.h
Comment (gggrace14) [reply velox/connectors/hive/HiveDataSink.h:545]: I see what Masha suggested was
> store indices of these in a member variable

Guess that should look more like the partitionChannels_ data member and might work better. Otherwise you still need to compute the indices of the child vectors of the data input every time you call makeDataInput(), as you're doing it right now.

**Review by gggrace14 — COMMENTED** [review]
@JkSelf The structure looks a lot better now and we're very close. Left some comments.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (JkSelf) [reply velox/connectors/hive/HiveDataSink.cpp]: Yes. Updated.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (JkSelf) [reply velox/connectors/hive/HiveDataSink.cpp]: @gggrace14 Support removing the partition columns with the bucket write.

File: velox/connectors/hive/HiveDataSink.h
Comment (JkSelf) [reply velox/connectors/hive/HiveDataSink.h]: @gggrace14 Added dataChannels_ member to store the indices of data columns.

**Comment by JkSelf** [comment]
@gggrace14 Thanks for your review. Can you help to review again? Thanks.

**Comment by JkSelf** [comment]
@mbasmanova Can you help to review again? Thanks.

**Comment by mbasmanova** [comment]
@gggrace14 Ge will help review and merge this PR.

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:408]:
```suggestion
auto columnIndex = getDataType(dataChannels_, inputType_)->getChildIdxIfExists(sortedProperty.at(i)->sortColumn()));
if (columnIndex.has_value()) {
    sortColumnIndices_.push_back(columnIndex.value());
}
```
Would the above work and be simpler?

File: velox/connectors/hive/HiveDataSink.cpp
Comment (gggrace14) [inline velox/connectors/hive/HiveDataSink.cpp:406]: Good catch here!

**Review by gggrace14 — COMMENTED** [review]
@JkSelf , thank you for revising! Left a bit more comments. Also I see this PR contains 7 commits. Would you be able to fold them into one? Not sure how they will end up in the master branch after merge.

**Comment by JkSelf** [comment]
@gggrace14 Resolved all your comments. Can you help to review again? Thanks.

**Review by gggrace14 — APPROVED** [review]
It looks great! Thank you for making the change, @JkSelf !

**Review by gggrace14 — APPROVED** [review]

File: velox/exec/tests/TableWriteTest.cpp
Comment (kewang1024) [inline velox/exec/tests/TableWriteTest.cpp:111]: Why is this utility function called ScanOutput?

**Review by kewang1024 — COMMENTED** [review]
Thanks @JkSelf for fixing the problem!

Overall PR lgtm, proposed some nits

File: velox/connectors/hive/HiveDataSink.cpp
Comment (JkSelf) [reply velox/connectors/hive/HiveDataSink.cpp]: It seems the Writer doesn't provide the API to get options here?

File: velox/exec/tests/TableWriteTest.cpp
Comment (JkSelf) [reply velox/exec/tests/TableWriteTest.cpp]: Updated to `getNonPartitionsColumns`.

**Comment by JkSelf** [comment]
@kewang1024 Resolved your comments. Can you help to review again? Thanks.

**Review by kewang1024 — APPROVED** [review]
Thank you @JkSelf, accepted with one NIT:
the functions can be static since they don't use any member variable (getNonPartitionTypes / getNonPartitionChannels / getNonPartitionsColumns / makeDataInput)

**Comment by gggrace14** [comment]
Hi @JkSelf , would you be able to address @kewang1024 's comment, squash the multiple commits and merge the PR? Let me know if you cannot merge the PR on your end.

**Comment by JkSelf** [comment]
@kewang1024 @gggrace14 Updated. Can you help to review again? Thanks.

**Review by gggrace14 — APPROVED** [review]

**Comment by gggrace14** [comment]
The revision looks good to me! Thank you, @JkSelf , for making the change! I'm merging the PR

**Pull request closed without merging**
